### PR TITLE
PERF-4314 Add merge results view for concurrent agent runs

### PR DESCRIPTION
**Pull request by tkmcmaster**
## Summary

Adds a "Merge Results" feature to both the PPaaS controller and the guide's results-viewer-react so operators can combine concurrent agent run results into a single unified view. Includes WASM OOM fixes, React rendering optimizations, request-count overview charts, and dependency updates applied opportunistically across the merge work.

## Reasoned Changes

<details>
<summary><b>Merge Results feature — controller (MergeSearchModal + TestResultsMerge + URL sync)</b></summary>

**What:** Adds a "Merge Results" workflow to the PPaaS controller: a `MergeSearchModal` that searches by S3 key path and fetches full `TestData` per result, a `TestResultsMerge` display component that renders merged charts and a per-agent breakdown, and `teststatus.tsx` URL query param wiring for `merge` (multi-value) and `mergeEndpoints` state.

**Why:** Concurrent load test agents each write a separate results file; without a merge view, operators have no single unified picture of a distributed run's combined throughput and latency. The controller merges by fetching each additional file's S3 content sequentially inside `onMergeLoad`, then calling `mergeResults()` to combine histograms at shared timestamps. A lazy-eviction strategy on `fetchedTests` (evict non-selected entries at the start of each new search) was chosen over immediate-on-deselect eviction to keep the implementation simple, accepting that the map grows transiently across search iterations.

</details>
<details>
<summary><b>Merge Results feature — guide ResultsViewer (TestResultsMerge + file-drop UX)</b></summary>

**What:** Adds a third "Merge Results" mode to the guide's `ResultsViewer`, accepting multiple file drops into a managed list, and a new `TestResultsMerge` component that parses all files concurrently, runs `mergeResults()`, shows a no-overlap warning, and offers a "Download Combined Results File" button.

**Why:** The guide results viewer serves local/offline use-cases where users have already downloaded per-agent files; giving them a drag-and-drop merge path avoids needing the controller at all. The download button serializes merged `DataPoint.rttHistogram` objects back to base64 using the new `toBase64()` WASM export, producing a file format round-trippable with the Rust binary. Storybook stories cover concurrent, staggered, no-overlap, and same-file scenarios so visual regressions in the merge rendering can be caught without a live dataset.

</details>
<details>
<summary><b>WASM memory safety — O(N²) fix and freeParsedEntries helper</b></summary>

**What:** Replaces the incremental `mergeAllDataPoints(...existing, ...dataPoints)` pattern (which cloned WASM histograms on every loop iteration) with an accumulate-then-merge pattern across all chart-grouping sites in both the controller and guide, and extracts a `freeParsedEntries` helper that centralises the WASM free-loops so no future results-clearing path can silently omit one.

**Why:** The old rest-param spread caused O(N²) WASM histogram clones for any endpoint with many tags, crashing the browser with an OOM when merging large result sets. The fix collects raw `DataPoint` references into `Map<string, DataPoint[]>` first (O(N) pointer copies, no WASM calls), then calls `mergeAllDataPoints` once per group. The `freeParsedEntries` helper is exported from `controller/components/TestResults/index.tsx` and consumed by `TestResultsMerge`; the guide has its own local equivalent that also frees `allParsed` source histograms after `mergeResults` clones what it needs.

</details>
<details>
<summary><b>toBase64() WASM export in hdr-histogram-wasm</b></summary>

**What:** Adds a `#[wasm_bindgen] pub fn to_base64()` method to `HDRHistogram` in `lib/hdr-histogram-wasm/src/lib.rs` using `V2Serializer`, serializing the histogram to the same binary format the Rust binary produces.

**Why:** The guide's "Download Combined Results File" button needs to write histogram data back to the JSON results format so the downloaded file can be re-loaded into any results viewer. `V2Serializer` was chosen because it matches the format the Rust pewpew binary uses in its own output, ensuring round-trip compatibility. The mock in `guide/results-viewer-react/__mocks__/hdr-histogram-wasm.js` returns a sentinel string to cover the download path in Jest tests.

</details>
<details>
<summary><b>Request Count overview charts added to TestResults and TestResultsCompare</b></summary>

**What:** Adds "Request Count by Endpoint", "Request Count by Host", and "Request Count by Agent/File" stacked-area charts to both the single-test `TestResults` view and the `TestResultsCompare` side-by-side view, with `Endpoint Data` collapsed by default.

**Why:** The per-agent throughput breakdown was the primary signal missing when reviewing multi-agent runs; adding it to the compare view lets users see whether two runs had similar load patterns before comparing latency. Collapsing `Endpoint Data` by default prevents upfront rendering of potentially hundreds of Chart.js canvases on large result sets. Pre-computing per-file request-count time series in `TestResultsCompare` as `useMemo` hooks moves WASM boundary crossings out of child component render paths.

</details>
<details>
<summary><b>Performance micro-optimizations and dependency updates</b></summary>

**What:** Wraps `FinalResultsTable` in `React.memo`, replaces the O(N) timestamp-join `dataPointsKey` with an O(1) count-first-last key, eliminates the intermediate flat array in `getSummaryData`, replaces `xlsx` with `write-excel-file`, bumps `js-yaml` to v5, and bumps CI tools (mdBook 0.5.3, wasm-pack 0.15.0).

**Why:** `FinalResultsTable` was re-rendering on every filter change even when `displayData` hadn't changed, causing jank with large result sets. The `xlsx` library carried large bundle weight and its `writeFile` API doesn't support the browser's streaming subpath; the new `exportResultsToExcel` wrapper accepts a dependency-injected `writeFn` so it is testable without a real browser. CI tool bumps fix the stale `rustwasm/wasm-pack` org URL and align with current release channels.

</details>

## Risk Assessment

Risk concentrates in the controller's sequential `onMergeLoad` fetch loop — each additional file is an awaited S3 round-trip with no loading indicator, and if `state.resultsData` is concurrently freed by a polling reload during those awaits, `mergeResults` will read freed WASM pointers. In practice this only affects **running** tests (Finished/Failed tests do not poll), and merge is primarily used on finished runs, so this is a low-probability edge case.

### Highest-Risk Areas

1. **controller/components/TestResults/index.tsx:1237–1287 (`onMergeLoad`)** — `allParsed[0]` is the live `state.resultsData` reference; a concurrent polling reload between `await fetchResults` calls could free those histograms before `mergeResults` reads them. Deferred: only affects running tests; restructuring to clone upfront would be expensive.

2. **controller/components/TestResults/index.tsx:1334–1357 (auto-merge `useEffect`)** — `loadMergeByTestIds` is defined inside the component body and captures `state` at definition time; if `state.resultsData` changes during a long sequential fetch, the function uses a stale results reference.

3. **controller/components/TestResults/MergeSearchModal.tsx:107–133** — `fetchedTests` is read inside a `useEffect` whose deps only include `searchResults`; stale `fetchedTests` could allow a redundant fetch to fire, but `pendingFetchIds` ref prevents true double-fetches, so this is safe in practice.

## Review Hotspots

1. **controller/components/TestResults/index.tsx:401–444 (`groupAndMergeByLabel` / `freeParsedEntries`)** — touched by three "Fix" commits; returned clones are freed inconsistently across call sites (some via `freeGroupedHistograms` in chart import callbacks, others via `useEffect` cleanup). A future call site that forgets the free will silently leak WASM heap.

2. **controller/components/TestResults/index.tsx:1334–1357 (auto-merge `useEffect`)** — `autoMergeTriggeredRef` stays `true` if `resultsData` resets (user switches results index), so auto-merge will not re-trigger for a new file even if `initialMergeTestIds` is still set in the URL.

## Testing

- Controller `TestResultsMerge` unit tests: `testresultsmerge.spec.tsx` (360 lines, 8 scenarios)
- Guide `TestResultsMerge` unit tests: `TestResultsMerge.test.tsx` (346 lines)
- Guide merge logic unit tests: `TestResults/__tests__/merge.test.ts` (213 lines)
- WASM mock updated with `toBase64()` stub; `freeParsedEntries` mock added to controller `TestResultsMerge` spec
- Storybook stories: concurrent agents, staggered start, no-overlap (warning), same-file-twice, different endpoints, empty state
- **Known gap:** `MergeSearchModal` has no unit tests — the search/fetch/toggle/eviction flow is Storybook-only

## Checklist

- [x] `./pr.sh` run successfully (builds everything, runs all tests including WASM rebuild)
- [x] WASM `toBase64()` requires rebuilding `hdr-histogram-wasm` with `wasm-pack build` before end-to-end guide testing
- [x] No breaking changes to existing TestResults/TestResultsCompare behavior
- [x] Controller `onMergeLoad` concurrent-reload risk (Risk #1) accepted — only affects running tests

### CODE REVIEW

File: guide/results-viewer-react/src/excelexport.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  { header: "Max",          cell: (row) => row.Max },
+  { header: "StdDev",       cell: (row) => row.StdDev },
+  { header: "Time",         cell: (row) => row.Time }
+];
+
+type WriteFn = (data: ExcelResultRow[], options: { columns: Column<ExcelResultRow>[]; sheet: string }) => { toFile: (fileName: string) => Promise<void> };
+
+export async function exportResultsToExcel (data: ExcelResultRow[], fileName: string, sheetName: string, writeFn: WriteFn = writeXlsxFile as unknown as WriteFn): Promise<void> {
+  const result = writeFn(data, { columns: RESULT_COLUMNS, sheet: sheetName });
+  await result.toFile(fileName);
```
[the remark this answers is not in the document]
Comment (tkmcmaster) [reply guide/results-viewer-react/src/excelexport.ts:42]: Investigated but this is a false positive. The library's `ReturnType` interface does expose `.toFile()`, and the correct option key is `columns` (not `schema`) — the library itself removed `schema` and replaced it with `columns`. Confirmed working via manual testing: the "Download as Excel" button produces a valid file with the `columns` key. No change needed.

File: controller/components/TestResults/index.tsx [hunk 2]
```diff
@@ -721,6 +995,8 @@ export const TestResults = React.memo(({ testData, initialResultsIndex, onResult
           compareTest: undefined,
           compareText: undefined,
           compareData: undefined,
+          mergedData: undefined,
+          mergedTestIds: [],
```
[the remark this answers is not in the document]
Comment (tkmcmaster) [reply controller/components/TestResults/index.tsx:1007]: Fixed — `mergedAgentTimeSeries: []` added to this reset path in commit 54efb16.

File: controller/components/TestResults/index.tsx [hunk 3]
```diff
@@ -760,6 +1037,8 @@ export const TestResults = React.memo(({ testData, initialResultsIndex, onResult
...
```
[the remark this answers is not in the document]
Comment (tkmcmaster) [reply controller/components/TestResults/index.tsx:1050]: Fixed — `mergedAgentTimeSeries: []` added to this reset path in commit 54efb16.

**Review by bryan-e-lopez — APPROVED** [review]
